elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/ElasticsearchFlowStage.scala

```diff
@@ -22,64 +22,191 @@ import akka.NotUsed
 import org.apache.http.message.BasicHeader
 import org.apache.http.util.EntityUtils
 
-object IncomingMessage {
+object IncomingIndexMessage {
   // Apply method to use when not using passThrough
   def apply[T](id: Option[String], source: T): IncomingMessage[T, NotUsed] =
-    IncomingMessage(id, source, NotUsed)
+    IncomingIndexMessage(id, source, NotUsed)
 
   // Apply method to use when not using passThrough
   def apply[T](id: Option[String], source: T, version: Long): IncomingMessage[T, NotUsed] =
-    IncomingMessage(id, source, NotUsed, Option(version))
+    IncomingIndexMessage(id, source, NotUsed, Option(version))
 
   // Java-api - without passThrough
   def create[T](id: String, source: T): IncomingMessage[T, NotUsed] =
-    IncomingMessage(Option(id), source)
+    IncomingIndexMessage(Option(id), source)
 
   // Java-api - without passThrough
   def create[T](id: String, source: T, version: Long): IncomingMessage[T, NotUsed] =
-    IncomingMessage(Option(id), source, version)
+    IncomingIndexMessage(Option(id), source, version)
 
   // Java-api - without passThrough
   def create[T](source: T): IncomingMessage[T, NotUsed] =
-    IncomingMessage(None, source)
+    IncomingIndexMessage(None, source)
 
   // Java-api - with passThrough
   def create[T, C](id: String, source: T, passThrough: C): IncomingMessage[T, C] =
-    IncomingMessage(Option(id), source, passThrough)
+    IncomingIndexMessage(Option(id), source, passThrough)
 
   // Java-api - with passThrough
   def create[T, C](id: String, source: T, passThrough: C, version: Long): IncomingMessage[T, C] =
-    IncomingMessage(Option(id), source, passThrough, Option(version))
+    IncomingIndexMessage(Option(id), source, passThrough, Option(version))
 
   // Java-api - with passThrough
   def create[T, C](id: String, source: T, passThrough: C, version: Long, indexName: String): IncomingMessage[T, C] =
-    IncomingMessage(Option(id), source, passThrough, Option(version), Option(indexName))
+    IncomingIndexMessage(Option(id), source, passThrough, Option(version), Option(indexName))
 
   // Java-api - with passThrough
   def create[T, C](source: T, passThrough: C): IncomingMessage[T, C] =
-    IncomingMessage(None, source, passThrough)
+    IncomingIndexMessage(None, source, passThrough)
 }
 
-final case class IncomingMessage[T, C](id: Option[String],
-                                       source: T,
-                                       passThrough: C,
-                                       version: Option[Long] = None,
-                                       indexName: Option[String] = None) {
+object IncomingUpdateMessage {
+  // Apply method to use when not using passThrough
+  def apply[T](id: String, source: T): IncomingMessage[T, NotUsed] =
```

Comment:
I think we should have factory `apply` and `create` methods only with the arguments that are mandatory. Then optional arguments can be set with `with***` builder methods. You can take a look at the [reference connector](https://github.com/akka/alpakka/blob/master/reference/src/main/scala/akka/stream/alpakka/reference/model.scala) which follows that convention.

Reply:
This would also bring elasticsearch connector one step closer to https://github.com/akka/alpakka/issues/1070